src/libclipper/src/containers.cpp

```diff
@@ -30,78 +34,191 @@ ModelContainer::ModelContainer(VersionedModelId model, int container_id,
+        old_info, static_cast<double>(processing_latency_micros));
+    processing_datapoints_[batch_size] = std::move(new_info);
   }
-  throughput_buffer_.push_back(new_throughput);
+
+  max_latency_ = std::max(processing_latency_micros, max_latency_);
+
+  EstimatorFittingThreadPool::submit_job(model_, replica_id_, [this]() {
+    try {
+      fit_estimator();
+    } catch (std::exception const &ex) {
+      log_error_formatted(LOGGING_TAG_CONTAINERS, "FITTING EXCEPTION: {}",
+                          ex.what());
+    }
+  });
 }
 
-void ModelContainer::set_batch_size(int batch_size) {
-  batch_size_ = batch_size;
+ModelContainer::LatencyInfo ModelContainer::update_mean_std(
+    LatencyInfo &info, double new_latency) {
+  double info_size = std::get<0>(info);
+  double mu = std::get<1>(info);
+  double std = std::get<2>(info);
+
+  double old_s = std::pow(std, 2) * info_size;
+
+  mu = ((mu * info_size) + new_latency) / (info_size + 1);
+  info_size += 1;
+
+  double new_s = old_s + ((info_size / std::max(1.0, info_size - 1)) *
+                          std::pow((mu - new_latency), 2));
+  std = std::sqrt(new_s / std::max(1.0, info_size));
+
+  return std::make_tuple(info_size, mu, std);
 }
 
-double ModelContainer::get_average_throughput_per_millisecond() {
-  boost::shared_lock<boost::shared_mutex> lock(throughput_mutex_);
-  return avg_throughput_per_milli_;
+void ModelContainer::set_batch_size(int batch_size) {
+  batch_size_ = batch_size;
 }
 
 size_t ModelContainer::get_batch_size(Deadline deadline) {
+  double budget =
+      static_cast<double>(std::chrono::duration_cast<std::chrono::microseconds>(
+                              deadline - std::chrono::system_clock::now())
+                              .count());
+  // Decay the provided latency budget by a pre-specified factor
+  // in order to provide enough slack for delivering the response
+  // to the user and/or coping with an anomolously high
+  // processing latency
+  budget = budget * budget_decay_;
+
   if (batch_size_ != DEFAULT_BATCH_SIZE) {
     return batch_size_;
   }
 
-  double current_time_millis =
-      std::chrono::duration_cast<std::chrono::milliseconds>(
-          std::chrono::system_clock::now().time_since_epoch())
-          .count();
-  double deadline_millis =
-      std::chrono::duration_cast<std::chrono::milliseconds>(
-          deadline.time_since_epoch())
-          .count();
-  double remaining_time_millis = deadline_millis - current_time_millis;
-  boost::shared_lock<boost::shared_mutex> lock(throughput_mutex_);
-  int batch_size =
-      static_cast<int>(avg_throughput_per_milli_ * remaining_time_millis);
-  if (batch_size < 1) {
-    batch_size = 1;
+  size_t curr_batch_size;
+  if (budget > static_cast<double>(max_latency_)) {
+    curr_batch_size = explore();
+  } else {
+    curr_batch_size = estimate(budget);
   }
-  return batch_size;
+  max_batch_size_ = std::max(curr_batch_size, max_batch_size_);
+  return curr_batch_size;
```

Comment:
Return an additional flag that indicates whether this batch size was selected using exploration or estimation.